File: src/rng.ts
/* ------------------------------------------------------------------ *
 * THE ONE SOURCE OF CHANCE
 *
 * Every random number the generator ever draws comes from here, and
 * the whole library is a pure function of the seed because of it. The
 * spec's word is "consistently": the seed varies detail, it does not
 * gamble on the outcome, and that promise is only checkable if there
 * is exactly one place chance enters.
 *
 * SplitMix32. Chosen over the xorshift32 the fn-11.1 placeholder used
 * because xorshift32's low bits are visibly correlated on neighbouring
 * seeds, and the panel has a reroll button - the owner will walk
 * consecutive seeds and must not be able to learn the sequence.
 * Math.random is not a candidate at any price: it is unseeded.
 * ------------------------------------------------------------------ */

export interface Rng {
  /** The next uniform in [0, 1). */
  next(): number;
  /** The next uniform in [min, max). */
  range(min: number, max: number): number;
}

/** A stream from `seed`. Any uint32 is a seed, zero included: the
 *  first thing SplitMix32 does is add the golden-ratio constant, so
 *  there is no degenerate state to special-case. */
export function createRng(seed: number): Rng {
  let state = seed >>> 0;

  const next = (): number => {
    state = (state + 0x9e_37_79_b9) >>> 0;
    let z = state;
    z = Math.imul(z ^ (z >>> 16), 0x21_f0_aa_ad);
    z = Math.imul(z ^ (z >>> 15), 0x73_5a_2d_97);
    z = (z ^ (z >>> 15)) >>> 0;
    // Divide by 2^32, never 2^32 - 1: the result has to exclude 1.
    return z / 0x1_00_00_00_00;
  };

  return {
    next,
    range: (min, max) => min + (max - min) * next(),
  };
}

File: crates/telperion-core/src/rng.rs
/// SplitMix32, including seed zero. Integer arithmetic is identical on native and Wasm.
#[derive(Debug, Clone)]
pub struct Rng {
    state: u32,
}
impl Rng {
    pub fn new(seed: u32) -> Self {
        Self { state: seed }
    }
    pub fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_add(0x9e3779b9);
        let mut z = self.state;
        z = (z ^ (z >> 16)).wrapping_mul(0x21f0aaad);
        z = (z ^ (z >> 15)).wrapping_mul(0x735a2d97);
        z ^ (z >> 15)
    }
    pub fn next_f64(&mut self) -> f64 {
        self.next_u32() as f64 / 4294967296.0
    }
    pub fn range(&mut self, min: f64, max: f64) -> f64 {
        min + (max - min) * self.next_f64()
    }
}
